packages/block-editor/src/components/link-control/index.js

```diff
@@ -0,0 +1,249 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+import { isFunction, noop, startsWith } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import {
+	Button,
+	ExternalLink,
+	Popover,
+} from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+
+import {
+	useCallback,
+	useState,
+	useEffect,
+	Fragment,
+} from '@wordpress/element';
+
+import {
+	safeDecodeURI,
+	filterURLForDisplay,
+	isURL,
+	prependHTTP,
+	getProtocol,
+} from '@wordpress/url';
+
+import { withInstanceId, compose } from '@wordpress/compose';
+import { withSelect } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import LinkControlSettingsDrawer from './settings-drawer';
+import LinkControlSearchItem from './search-item';
+import LinkControlSearchInput from './search-input';
+
+function LinkControl( {
+	className,
+	currentLink,
+	currentSettings,
+	fetchSearchSuggestions,
+	instanceId,
+	onClose = noop,
+	onKeyDown = noop,
+	onKeyPress = noop,
+	onLinkChange = noop,
+	onSettingsChange = { noop },
+} ) {
+	// State
+	const [ inputValue, setInputValue ] = useState( '' );
+	const [ isEditingLink, setIsEditingLink ] = useState( false );
+
+	// Effects
+	useEffect( () => {
+		// If we have a link then stop editing mode
+		if ( currentLink ) {
+			setIsEditingLink( false );
+		} else {
+			setIsEditingLink( true );
+		}
+	}, [ currentLink ] );
+
+	// Handlers
+
+	/**
+	 * onChange LinkControlSearchInput event handler
+	 *
+	 * @param {string} value Current value returned by the search.
+	 */
+	const onInputChange = ( value = '' ) => {
+		setInputValue( value );
+	};
+
+	// Utils
+	const startEditMode = () => {
+		if ( isFunction( onLinkChange ) ) {
+			onLinkChange();
+		}
+	};
+
+	const closeLinkUI = () => {
+		resetInput();
+		onClose();
+	};
+
+	const resetInput = () => {
+		setInputValue( '' );
+	};
+
+	const handleDirectEntry = ( value ) => {
+		let type = 'URL';
+
+		const protocol = getProtocol( value ) || '';
+
+		if ( protocol.includes( 'mailto' ) ) {
+			type = 'mailto';
+		}
+
+		if ( protocol.includes( 'tel' ) ) {
+			type = 'tel';
+		}
+
+		if ( startsWith( value, '#' ) ) {
+			type = 'internal';
+		}
+
+		return Promise.resolve(
+			[ {
+				id: '-1',
+				title: value,
+				url: type === 'URL' ? prependHTTP( value ) : value,
+				type,
```

Comment:
Was there any planned usage for the various `type`s associated with manual entry? As part of the ongoing discussion in #18061, I've been considering whether this is something where the explicit absence of the type could be a reasonably good indicator on its own for marking a value as "manual entry". The sorts of things that someone might want to do with `type` should be equally achievable with the `@wordpress/url` utilities, if a developer were so inclined. One of the issues I encountered in #19827 was trying to decide what the best complete set of these "subtypes" should be, where we would probably be better off to just not have to answer that.

Reply:
> Was there any planned usage for the various `type`s associated with manual entry? As part of the ongoing discussion in #18061, I've been considering whether this is something where the explicit absence of the type could be a reasonably good indicator on its own for marking a value as "manual entry". The sorts of things that someone might want to do with `type` should be equally achievable with the `@wordpress/url` utilities, if a developer were so inclined. One of the issues I encountered in #19827 was trying to decide what the best complete set of these "subtypes" should be, where we would probably be better off to just not have to answer that.

Follow-up: #20051